### PR TITLE
test(redteam): complete_overdue_exam_session attack-surface spec (#559, #560, #561, #562)

### LINKED ISSUE #559: red-team AX: complete_overdue_exam_session IDOR — foreign session_id
## Vector AX (HIGH)

**RPC:** complete_overdue_exam_session(p_session_id uuid)
**Commit:** 5cbbfe8

### Attack
Student-B authenticates and calls complete_overdue_exam_session(session_id_owned_by_A). The RPC must raise 'session not found or not accessible' because the FOR UPDATE query requires qs.student_id = auth.uid(). No spec currently exercises this ownership guard.

### Spec sketch
In a new spec file rpc-complete-overdue-exam.spec.ts:
1. Create student-A session with mode='mock_exam', past deadline.
2. Sign in as student-B.
3. Call complete_overdue_exam_session(a_session_id) directly via createAuthenticatedClient.
4. Expect RAISE EXCEPTION 'session not found or not accessible'.

### Notes
Mechanically identical to Vector AO (complete_empty_exam_session IDOR, also unspecced).

**Label:** testing

### LINKED ISSUE #560: red-team AY: complete_overdue_exam_session called on non-overdue session
## Vector AY (HIGH)

**RPC:** complete_overdue_exam_session(p_session_id uuid)
**Commit:** 5cbbfe8

### Attack
Student calls complete_overdue_exam_session on their own active, still-running session (deadline not yet passed). The RPC must raise 'session is not overdue' per defense-in-depth invariant. No spec exercises this guard. If the guard failed, a student could force-close their own exam early with partial answers scored as a pass.

### Spec sketch
In rpc-complete-overdue-exam.spec.ts:
1. Create student's own session with mode='mock_exam', started_at = now(), time_limit_seconds = 3600 (far future deadline).
2. Call complete_overdue_exam_session(session_id) directly.
3. Expect RAISE EXCEPTION 'session is not overdue'.

### Notes
This is the primary integrity guard for the new RPC. Highest-priority test case.

**Label:** testing

### LINKED ISSUE #561: red-team AZ: complete_overdue_exam_session mode confusion — study session
## Vector AZ (HIGH)

**RPC:** complete_overdue_exam_session(p_session_id uuid)
**Commit:** 5cbbfe8

### Attack
Student calls complete_overdue_exam_session on a study-mode or quick_quiz session_id. The RPC must raise 'session is not a mock exam'. No spec exercises this mode guard. Without it, a student could force-score a study session as an exam.

### Spec sketch
In rpc-complete-overdue-exam.spec.ts:
1. Create student's own session with mode='study' (or 'quick_quiz'), past deadline.
2. Call complete_overdue_exam_session(session_id) directly.
3. Expect RAISE EXCEPTION 'session is not a mock exam'.

### Notes
Mechanically identical to Vector AP (complete_empty_exam_session mode confusion, also unspecced).

**Label:** testing

### LINKED ISSUE #562: red-team BC: start_exam_session + concurrent complete_overdue race on same overdue session
## Vector BC (MEDIUM)

**RPCs:** start_exam_session, complete_overdue_exam_session
**Commit:** 5cbbfe8

### Attack
Student's old session is overdue. Student calls start_exam_session (which internally calls PERFORM complete_overdue_exam_session) at the same time a second tab directly calls complete_overdue_exam_session on the same overdue session_id. Both paths enter complete_overdue_exam_session concurrently.

### Expected defence
FOR UPDATE lock inside complete_overdue_exam_session serialises both callers. The second caller sees v_ended_at IS NOT NULL and returns the idempotent payload (real stored score). No double-count, no data corruption.

### Spec sketch
In rpc-complete-overdue-exam.spec.ts:
1. Create overdue mock_exam session.
2. Fire complete_overdue_exam_session and start_exam_session concurrently via Promise.all.
3. Assert exactly one audit event 'exam.expired' is written.
4. Assert quiz_sessions row has ended_at set exactly once.
5. Assert start_exam_session returns a new session_id (different from old).

### Notes
This pins the FOR UPDATE concurrency guarantee. Low risk in practice but worth a pinning spec given the idempotency branch was fixed vs AQ.

**Label:** testing

### PR DESCRIPTION
**Pull request by okpilot**
## What

New red-team Playwright spec covering the `complete_overdue_exam_session` SECURITY DEFINER RPC, plus the attack-surface matrix flip. **Coverage-only — no production code change.**

## Vectors covered (attack-surface.md GAP → COVERED)

| Vector | Attack | Expected defense |
|--------|--------|------------------|
| **AX** (HIGH) | different student completes a foreign session (IDOR) | `session not found or not accessible` (ownership + org scope in `FOR UPDATE`) |
| **AY** (HIGH) | owner completes a not-yet-overdue session | `session is not overdue` (deadline + 30s grace invariant) |
| **AZ** (HIGH) | owner completes a non-exam (`quick_quiz`) session | `session is not an exam` (mode guard) |
| **BB** (LOW) | owner completes a soft-deleted session | `session not found or not accessible` (`deleted_at IS NULL`) |
| **BC** (MED) | concurrent completion of the same overdue session | `FOR UPDATE` serialises; second caller is idempotent |

Plus: soft-deleted-user guard (`user not found or inactive`) and positive paths for both `mock_exam` and `internal_exam`.

## Verification

- Assertions verified against the latest RPC definition (mig `20260429000008`); guard order confirmed by impl-critic + semantic-reviewer.
- biome / tsc / 3497 Vitest / build all green locally. CodeRabbit local: no findings.
- The redteam Playwright suite runs in CI (`redteam.yml`) — local stack not available.

## Notes

- Unauthenticated-call coverage for this RPC is intentionally deferred to the Hub A `server-action-unauthenticated.spec.ts` (canonical home for unauth cases).

Closes #559
Closes #560
Closes #561
Closes #562

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Added a comprehensive end-to-end test suite covering exam session completion scenarios, authorization checks, edge cases, concurrent calls, and cleanup behaviors.

* **Documentation**
  * Updated internal testing and security documentation to reflect new coverage, tracking guidance, and positive-pattern learnings.

### CODE REVIEW

File: apps/web/e2e/redteam/rpc-complete-overdue-exam.spec.ts [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    expect(error?.message ?? '').toMatch(/session not found or not accessible/i)
+    expect(data).toBeNull()
+
+    // The session must remain active — the attacker did not complete it.
+    const { data: row } = await admin
+      .from('quiz_sessions')
+      .select('ended_at')
+      .eq('id', sessionId)
+      .single()
+    expect(row?.ended_at).toBeNull()
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/e2e/redteam/rpc-complete-overdue-exam.spec.ts]: Fixed in e01bbccc — all four `admin.from('quiz_sessions').select(...).single()` assertion reads now destructure `{ error: readErr }` and `expect(readErr).toBeNull()` before consuming the row, per code-style.md §5 (and the open test/e2e sweep #722). Thanks!
